### PR TITLE
OPSEXP-1046 tomcat base multistage builds

### PR DESCRIPTION
**Pull request by alxgomz**
create a matrix of OS/tomcat versions/ java images for evalution purposes
Moves build process to github action & changes naming scheme as described in README.md:

* STARTS CREATING NEW TAGS 
* STOPS CREATING TAGS WITH PREVIOUS NAMING (children images will need to align)

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
+name: Alfresco tomcat base Docker image
+on:
+  push:
+    branches:
+      #TODO add master when confident
+      - opsexp-[0-9]+**
+
+env:
+  IMAGE_REGISTRY_NAMESPACE: alxgomz
```
Comment (Mieszko96) [inline .github/workflows/main.yml:9]: Do we want to use your nickname as registry namespace?

File: Dockerfile [hunk 2]
```diff
@@ -1,73 +1,139 @@
...
 # Alfresco Base Tomcat Image
 # see also https://github.com/docker-library/tomcat
+ARG JDIST
 ARG JAVA_MAJOR
 ARG DISTRIB_NAME
 ARG DISTRIB_MAJOR
 ARG TOMCAT_MAJOR
 
-FROM java-$JAVA_MAJOR-$DISTRIB_NAME-$DISTRIB_MAJOR AS tomcat-8
+FROM quay.io/alfresco/alfresco-base-java:$JDIST${JAVA_MAJOR}-$DISTRIB_NAME${DISTRIB_MAJOR}-opsexp-1046-java-base-msb AS debian11
```
Comment (gionn) [inline Dockerfile:10]:
```suggestion
FROM quay.io/alfresco/alfresco-base-java:${JDIST}${JAVA_MAJOR}-${DISTRIB_NAME}${DISTRIB_MAJOR}-opsexp-1046-java-base-msb AS debian11
```

for consistency

File: .github/workflows/main.yml (hunk 1 above)
Comment (alxgomz) [reply .github/workflows/main.yml:9]: oops. fixed

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        uses: actions/checkout@v2
+      - id: vars
+        name: compute image tag
+        env:
+          IMAGE_BASE_NAME: tomcat${{ matrix.tomcat_major }}-${{ matrix.jdist }}${{ matrix.java_major }}-${{ matrix.base_image.flavor }}${{ matrix.base_image.major }}
+        run: |
+          if [[ "${{ github.ref_name }}" != "master" ]]
+          then
+            echo "::set-output name=image_tag::${IMAGE_BASE_NAME}-${GITHUB_REF_NAME//\//-}"
+          elif [[ "${{ github.event.head_commit.message }}" =~ .*\[release\].* ]]
```
Comment (gionn) [inline .github/workflows/main.yml:57]: not sure if it's a good idea to do an explicit release management for base image, is this working as before?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            quay.io/${{ env.IMAGE_REGISTRY_NAMESPACE }}/${{ env.IMAGE_REPOSITORY }}:${{ steps.vars.outputs.image_tag }}
+          target: TOMCAT_BASE_IMAGE
+      - id: dh_push
+        if: contains(github.event.head_commit.message, '[release]') && github.ref_name == 'master'
+        name: Build and push to Docker Hub
+        uses: docker/build-push-action@v2.8.0
+        with:
+          push: true
+          build-args: |
+            no-cache=true
```
Comment (gionn) [inline .github/workflows/main.yml:104]: are you forcing a full rebuild when also releasing on dockerhub? why?

File: .github/workflows/main.yml (hunk 3 above)
Comment (alxgomz) [reply .github/workflows/main.yml:57]: if referring to the "[release] flag in the commit message, yes this is how it's done today

File: .github/workflows/main.yml (hunk 4 above)
Comment (alxgomz) [reply .github/workflows/main.yml:104]: not particularly. Just reusing the --no-cache flag that was already used in previous build, but I have to say that it sounds safer to use it given the dynamic nature of this builds

Comment (gionn) [reply .github/workflows/main.yml:104]: and I see that builds run in a couple of minutes so nevermind

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
+name: Alfresco tomcat base Docker image
+on:
+  push:
+    branches:
+      - opsexp-[0-9]+**
+      - master
```
Comment (gionn) [inline .github/workflows/main.yml:6]: need to add dependabot

File: Dockerfile [hunk 6]
```diff
@@ -1,73 +1,139 @@
...
+	#fi && \
+        tar -zxf libapr.tar.gz --strip-components=1 -C ${BUILD_DIR}/libapr; \
+	# signing keys seem unavailable.... \
+	#if gpg --batch --verify apr-util.tar.gz.asc apr-util.tar.gz; then \
+	#	echo signature checked; \
+	#else \
+	#	keyID=$(gpg --batch --verify apr-util.tar.gz.asc apr-util.tar.gz 2>&1 | awk '/RSA\ /{print $NF}'); \
+	#	gpg --server hkp://keys.gnupg.net --recv-keys "$keyID"; \
+	#	gpg --batch --verify apr-util.tar.gz.asc apr-util.tar.gz; \
+	#fi && \
```
Comment (gionn) [inline Dockerfile:104]: is signatures check not working worth investigating?

otherwise I would remove all the commented lines

File: Dockerfile [hunk 7]
```diff
@@ -90,69 +154,63 @@ RUN set -eux; \
...
 	# Replace default 404,403,500 page
-	sed -i "$ d" /usr/local/tomcat/conf/web.xml ; \
-	sed -i -e "\$a\    <error-page\>\n\        <error-code\>404<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\    <error-page\>\n\        <error-code\>403<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\    <error-page\>\n\        <error-code\>500<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\n\<\/web-app\>" /usr/local/tomcat/conf/web.xml ; \
-	yum clean all
+	sed -i "$ d" conf/web.xml ; \
+	sed -i -e "\$a\    <error-page\>\n\        <error-code\>404<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\    <error-page\>\n\        <error-code\>403<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\    <error-page\>\n\        <error-code\>500<\/error-code\>\n\        <location\>\/error.jsp<\/location\>\n\    <\/error-page\>\n\n\<\/web-app\>" conf/web.xml 
 
+FROM quay.io/alfresco/alfresco-base-java:$JDIST${JAVA_MAJOR}-$DISTRIB_NAME${DISTRIB_MAJOR} AS TOMCAT_BASE_IMAGE
```
Comment (gionn) [inline Dockerfile:185]:
```suggestion
FROM quay.io/alfresco/alfresco-base-java:$JDIST${JAVA_MAJOR}-$DISTRIB_NAME${DISTRIB_MAJOR}
```

alias is not needed if it's the last stage

Comment (alxgomz) [reply Dockerfile:185]: this is used by the `--target` argument of docker build

File: Dockerfile [hunk 8]
```diff
@@ -90,69 +154,63 @@ RUN set -eux; \
...
-	chmod 777 logs work ; \
+	chmod 330 logs work ; \
 	\
 	# Security improvements:
 	# Remove server banner, Turn off loggin by the VersionLoggerListener, enable remoteIP valve so we know who we're talking to
 	sed -i \
           -e "s/\  <Listener\ className=\"org.apache.catalina.startup.VersionLoggerListener\"/\  <Listener\ className=\"org.apache.catalina.startup.VersionLoggerListener\"\ logArgs=\"false\"/g" \
           -e "s%\(^\s*</Host>\)%\t<Valve className=\"org.apache.catalina.valves.RemoteIpValve\" />\n\n\1%" \
-          -e "s/\    <Connector\ port=\"8080\"\ protocol=\"HTTP\/1.1\"/\    <Connector\ port=\"8080\"\ protocol=\"HTTP\/1.1\"\n\               Server=\" \"/g" /usr/local/tomcat/conf/server.xml; \
+          -e "s/\    <Connector\ port=\"8080\"\ protocol=\"HTTP\/1.1\"/\    <Connector\ port=\"8080\"\ protocol=\"HTTP\/1.1\"\n\               Server=\" \"/g" conf/server.xml; \
```
Comment (Mieszko96) [inline Dockerfile:176]: Out of curiosity, why we want to use relative path?

**Review by gionn — APPROVED** [review]

File: Dockerfile (hunk 8 above)
Comment (alxgomz) [reply Dockerfile:176]: 2 things:
* multistage build makes the installtion path irrelevant (we're actually unzipping in /build/something) 
* we're using WORKDIR just before so we know that commands runs from within the tomcat distribution root dir)

**Pull request merged**